pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -853,6 +857,37 @@ type Artifact struct {
 	FromExpression string `json:"fromExpression,omitempty" protobuf:"bytes,11,opt,name=fromExpression"`
 }
 
+// CleanPath validates and cleans the artifact path.
+func (a *Artifact) CleanPath() error {
+	if a.Path == "" {
+		return argoerrs.InternalErrorf("Artifact '%s' did not specify a path", a.Name)
+	}
+
+	// Ensure that the artifact path does not use directory traversal to escape a
+	// "safe" sub-directory, assuming malicious user input is present. For example:
+	// inputs:
+	//    artifacts:
+	//      - name: a1
+	//        path: /tmp/safe/{{ inputs.parameters.user-input }}
+	//
+	// Any resolved path should always be within the /tmp/safe/ directory.
+	safeDir := ""
```

Comment:
We should not use regex for this, they are very brittle, and sometimes create un-fixable bugs. This is demonstrated by the many edge case bug we have when they were used in the past.

Could we use `path.Clean`?

https://dzx.cz/2021/04/02/go_path_traversal/

Reply:
Yes, `path.Clean` is used a few lines below. The "safeDir" is the best-effort guess as to what the template creator was attempting to do. See [this comment](https://github.com/argoproj/argo-workflows/pull/7187#issuecomment-970415218) above for a table of use cases, and the full suite of [test cases](https://github.com/argoproj/argo-workflows/pull/7187/files#diff-ebcb4d1a2cd98dee1d7ec51113c3dee40128e21e2d259699163449f34c431e98R83) that precisely show how this will behave.

Reply:
I think we should probably be cleaning at the point of usage, not before. I.e. before we invoke LoadArtifact and SaveArtifact. Thoughts?

Reply:
The only usages outside `LoadArtifact` and `SaveArtifact` are in `workflowpod.addInputArtifactsVolumes` and `validate.validateInputs`. In those places, there were previously errors raised if the artifact path was empty, so I thought it would also be prudent to further validate the given path by way of cleaning it.

Since `path.Clean` does not mutate the final destination from an OS perspective (i.e. `/tmp/../tmp` == `/tmp`), cleaning isn't strictly necessary. This method could be renamed `ValidatePath`, and remain just as useful.

I don't see any harm in this implementation as-is, since `path.Clean` should be idempotent. I'm open to other opinions, if someone has a better suggestion.